File: task1/py-deepimpute/run_model.py
#!/usr/bin/env python

import argparse
from imputation_model import imputation_model
from multiprocessing import Pool
import os
import glob
import random


def get_args():
    """Set up command-line interface and get arguments."""
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--input_dir", type=str, required=True)
    parser.add_argument("-o", "--output_dir", type=str, required=True)
    return parser.parse_args()


def set_batches(orig_list, batch_size, shuffle=False):
    """Split large list into smaller batches."""
    if shuffle:
        random.shuffle(orig_list)
    for i in range(0, len(orig_list), batch_size):
        yield orig_list[i:i + batch_size]


def get_filenames(input_dir, pattern="*.csv"):
    """Retrieve filenames without extensions of all downsampled files."""
    input_files = [os.path.basename(x) for x in glob.glob(
        os.path.join(input_dir, pattern))]
    # get filenames without extensions
    filenames = [os.path.splitext(f)[0] for f in input_files]

    return filenames


def main():
    """Main function."""
    args = get_args()
    input_dir = args.input_dir
    output_dir = args.output_dir
    # be careful not to exceed memory limit (240g)
    nworkers = 30  # used to impute files in parallel
    ncores = 6  # used to run model on single file

    # get filenames without extensions
    filenames = get_filenames(input_dir, "*.csv")
    # if needed, split into smaller batches to reduce memory usage
    batches = set_batches(filenames, 30, shuffle=True)

    # set input and output file paths
    for batch in batches:
        input_paths = [f"{input_dir}/{b}.csv" for b in batch]
        output_paths = [f"{output_dir}/{b}_imputed.csv" for b in batch]

        model_inputs = zip(input_paths, output_paths,
                           len(filenames)*[int(ncores)])

        # run model in parallel
        with Pool(processes=int(nworkers)) as pool:
            pool.map(imputation_model, model_inputs)


if __name__ == "__main__":
    main()

File: task1/r-magic/run_model.R
#!/usr/bin/env Rscript
Sys.setenv("RETICULATE_MINICONDA_ENABLED" = FALSE)
suppressPackageStartupMessages({
  library("data.table")
  library("tibble")
  library("Rmagic")
})

# Read arguments
args <- commandArgs(trailingOnly = TRUE)
input_dir <- args[1]
output_dir <- args[2]

# Set cores if needed
# Do not use more than 30 cores - it might slow down the performance
ncores <- 30

# Load your model
source("/imputation_model.R")

# Get input files
input_filenames <- list.files(input_dir, pattern = "*.csv")
# Retrieve the filenames without ext of all downsampled data
filenames <- tools::file_path_sans_ext(input_filenames)
# Add "_imputed" to create prediction/output filenames
output_filenames <- paste0(filenames, "_imputed.csv")

# split into smaller chunks (index) to reduce memory usage
set.seed(1234)
chunks <- split(
  sample(seq_along(input_filenames)),
  cut(seq_along(input_filenames), 3, labels = FALSE)
)

for (c in chunks) {
  parallel::mclapply(c, function(i) {
    # read input data
    input_path <- file.path(input_dir, input_filenames[i])
    input_data <- data.table::fread(input_path, data.table = FALSE)
    input_data <- tibble::column_to_rownames(input_data, "V1")

    # apply your model for imputation
    pred_data <- imputation_model(input_data)

    # write out the imputed data
    data.table::fwrite(
      data.table::as.data.table(pred_data, keep.rownames = ""),
      file.path(output_dir, output_filenames[i])
    )
  }, mc.cores = ncores)
}
